src/backend/distributed/operations/shard_cleaner.c

```diff
@@ -274,15 +642,310 @@ TryDropShard(GroupShardPlacement *placement)
+												   &result);
+	if (queryResult != RESPONSE_OKAY || !IsResponseOK(result) || PQntuples(result) != 1 ||
+		PQnfields(result) != 1)
+	{
+		ReportResultError(connection, result, ERROR);
+	}
+
+	operationdId = SafeStringToUint64(PQgetvalue(result, 0, 0 /* nodeId column*/));
+	CloseConnection(connection);
+
+	return operationdId;
+}
+
+
+/*
+ * ListCleanupRecords lists all the current cleanup records.
+ */
+static List *
+ListCleanupRecords(void)
+{
+	Relation pgDistCleanup = table_open(DistCleanupRelationId(), AccessShareLock);
+	TupleDesc tupleDescriptor = RelationGetDescr(pgDistCleanup);
+
+	List *recordList = NIL;
+	int scanKeyCount = 0;
+	bool indexOK = false;
+
+	SysScanDesc scanDescriptor = systable_beginscan(pgDistCleanup, InvalidOid,
+													indexOK, NULL, scanKeyCount, NULL);
+
+	HeapTuple heapTuple = NULL;
+	while (HeapTupleIsValid(heapTuple = systable_getnext(scanDescriptor)))
+	{
+		CleanupRecord *record = TupleToCleanupRecord(heapTuple, tupleDescriptor);
+		recordList = lappend(recordList, record);
+	}
+
+	systable_endscan(scanDescriptor);
+	table_close(pgDistCleanup, NoLock);
+
+	return recordList;
+}
+
+
+/*
+ * ListCleanupRecordsForCurrentOperation lists all the cleanup records for
+ * current operation.
+ */
+static List *
+ListCleanupRecordsForCurrentOperation(void)
+{
+	/* We must have a valid OperationId. Any operation requring cleanup
+	 * will call StartNewOperationNeedingCleanup.
+	 */
+	Assert(CurrentOperationId != INVALID_OPERATION_ID);
+
+	Relation pgDistCleanup = table_open(DistCleanupRelationId(), AccessShareLock);
+	TupleDesc tupleDescriptor = RelationGetDescr(pgDistCleanup);
+
+	ScanKeyData scanKey[1];
+	ScanKeyInit(&scanKey[0], Anum_pg_dist_cleanup_operation_id, BTEqualStrategyNumber,
+				F_INT8EQ, UInt64GetDatum(CurrentOperationId));
+
+	int scanKeyCount = 1;
+	Oid scanIndexId = InvalidOid;
+	bool useIndex = false;
+	SysScanDesc scanDescriptor = systable_beginscan(pgDistCleanup, scanIndexId, useIndex,
+													NULL,
+													scanKeyCount, scanKey);
+
+	HeapTuple heapTuple = NULL;
+	List *recordList = NIL;
+	while (HeapTupleIsValid(heapTuple = systable_getnext(scanDescriptor)))
+	{
+		CleanupRecord *record = TupleToCleanupRecord(heapTuple, tupleDescriptor);
+		recordList = lappend(recordList, record);
+	}
+
+	systable_endscan(scanDescriptor);
+	table_close(pgDistCleanup, NoLock);
+
+	return recordList;
+}
+
+
+/*
+ * TupleToCleanupRecord converts a pg_dist_cleanup record tuple into a CleanupRecord struct.
+ */
+static CleanupRecord *
+TupleToCleanupRecord(HeapTuple heapTuple, TupleDesc tupleDescriptor)
+{
+	Datum datumArray[Natts_pg_dist_cleanup];
+	bool isNullArray[Natts_pg_dist_cleanup];
+	heap_deform_tuple(heapTuple, tupleDescriptor, datumArray, isNullArray);
+
+	CleanupRecord *record = palloc0(sizeof(CleanupRecord));
+
+	record->recordId =
+		DatumGetUInt64(datumArray[Anum_pg_dist_cleanup_record_id - 1]);
+
```

Comment:
missing record->operationId

this is causing the issues, since operationId is set to 0 and we can always get the lock for 0 in the maintenance daemon

Reply:
Thanks for catching this @marcocitus. This worried me a bit so I have added a deterministic isolation test that will fail without the fix.